### ISSUE TITLE
fix: `BlocBuilder` builder callback called only once even if emit was called twice in BLoC handler.

### ISSUE THREAD

**Issue by comatory** [felangel/bloc#3197]
**Description**

When I have a async handler for an event and that handler calls `emit` twice right after each other, it seems like only the final emitted state is received by `BlocBuilder`.
If there is some kind of delay with `await`, then the behaviour is correct and state is emitted for each `emit` call.

**Steps To Reproduce**

1. Go to [my gist](https://gist.github.com/comatory/c1bd789802312b4d287fee40e0f939de) that contains reproducible example.
2. Create new Flutter app, run the application in development mode
3. Uncheck the checkbox, tap "Clear"
4. Tap start button
5. Look into console and compare it with printed logs on screen

**Expected Behavior**

If I call `emit` twice after each other, I would expect `BlocBuilder` builder callback to be called twice, first with `status=loading` and then `status=initial` values.

This can be observed with console, with unchecked checkbox the fake request only shows:

```
I/flutter (12532): NetworkStatus.initial
```

But the attached listener in `initState` which is logging each state change displays:

```
NetworkStatus.loading
NetworkStatus.initial
```

The `_mapNetworkStartEventToState` handler is async, if I place `await Future.delayed(Duration.zero)` below `!state.allow` condition, it works as expected. It looks like the updates to `BlocBuilder` are somehow batched. Can this be configured somehow?

**Screenshots**
![Screenshot_20220204-115723](https://user-images.githubusercontent.com/5990424/152517659-7f48fc1f-3b2e-4aa5-9634-b3499f19d2f4.jpg)

**Additional Context**

```
[✓] Flutter (Channel stable, 2.8.1, on macOS 11.6.1 20G224 darwin-x64, locale xx-XX)
    • Flutter version 2.8.1 at /Users/user/fvm/versions/2.8.1
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 77d935af4d (7 weeks ago), 2021-12-16 08:37:33 -0800
    • Engine revision 890a5fca2e
    • Dart version 2.15.1

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/user/Library/Android/sdk
    • Platform android-31, build-tools 31.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7281165)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2020.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7281165)

[✓] IntelliJ IDEA Community Edition (version 2020.2.1)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.58.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension can be installed from:
      🔨 https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter
```

`pubspec.lock`:
```
  flutter_bloc:
    dependency: "direct main"
    description:
      name: flutter_bloc
      url: "https://pub.dartlang.org"
    source: hosted
    version: "8.0.1"
```

**Comment by felangel** [comment]
Hi @comatory 👋
Thanks for opening an issue!

I believe this is a duplicate of #2260. Let me know if that helps 👍

**Comment by comatory** [comment]
> Hi @comatory 👋 Thanks for opening an issue!
> 
> I believe this is a duplicate of #2260. Let me know if that helps 👍

Yes that looks like it and I read the replies:

> This is expected because BlocBuilder can only rebuild at 60fps. In this case you are emitting states faster than BlocBuilder can rebuild.

> You can reproduce the same behavior with setState

Mmm, whaat? 😮 

Do I understand correctly that this is a limitation of Flutter?

**Comment by felangel** [comment]
> > Hi @comatory 👋 Thanks for opening an issue!
> > I believe this is a duplicate of #2260. Let me know if that helps 👍
> 
> Yes that looks like it and I read the replies:
> 
> > This is expected because BlocBuilder can only rebuild at 60fps. In this case you are emitting states faster than BlocBuilder can rebuild.
> 
> > You can reproduce the same behavior with setState
> 
> Mmm, whaat? 😮
> 
> Do I understand correctly that this is a limitation of Flutter?

If you emit more than 60fps then you can't render a frame for each state. Instead you should aggregate the data if all pieces of data are critical or you can throttle the bloc to only emit at 60fps.

**Comment by comatory** [comment]
> If you emit at a frequency that is more than 60fps then you can't render a frame for each state. Instead you should aggregate the data if all pieces of data are critical or you can throttle the bloc to only emit at 60fps.

OK but is this a limitation of Flutter framework or just way this lib was designed?

Aggregating data seems like a technical detail I don't really want to deal with. How would I throttle the bloc?

**Comment by felangel** [comment]
> > If you emit at a frequency that is more than 60fps then you can't render a frame for each state. Instead you should aggregate the data if all pieces of data are critical or you can throttle the bloc to only emit at 60fps.
> 
> OK but is this a limitation of Flutter framework or just way this lib was designed?
> 
> Aggregating data seems like a technical detail I don't really want to deal with. How would I throttle the bloc?

You can provide a custom event transformer and apply a throttle from something like package:rxdart or package:stream_transform.

**Comment by felangel** [comment]
Closing for now but feel free to comment with any additional questions and I'm happy to continue the conversation 👍
